log/sequencer.go

```diff
@@ -230,12 +237,60 @@ func (s Sequencer) initMerkleTreeFromStorage(ctx context.Context, currentRoot tr
 	return s.buildMerkleTreeFromStorageAtRoot(ctx, currentRoot, tx)
 }
 
-// SequenceBatch wraps up all the operations needed to take a batch of queued leaves
-// and integrate them into the tree.
-// TODO(Martin2112): Can possibly improve by deferring a function that attempts to rollback,
-// which will fail if the tx was committed. Should only do this if we can hide the details of
-// the underlying storage transactions and it doesn't create other problems.
-func (s Sequencer) SequenceBatch(ctx context.Context, logID int64, limit int, guardWindow, maxRootDurationInterval time.Duration) (int, error) {
+// sequencingTask is used by Sequencer to extract (pre-)sequenced LogLeaf
+// entries from the storage and commit to the ordering.
+type sequencingTask interface {
+	// load is invoked once to mine a batch of sequenced entries.
+	load(ctx context.Context, limit int, cutoff time.Time) ([]*trillian.LogLeaf, error)
+	// store is invoked once to send the batch back for potential storage update.
+	store(ctx context.Context, leaves []*trillian.LogLeaf) error
+}
+
+// logSequencingTask is a sequencingTask implementation for Log mode.
+// TODO(pavelkalinnikov): Implement it for Sequenced Log mode similarly.
+type logSequencingTask struct {
+	label      string
+	treeSize   int64
+	timeSource util.TimeSource
+	dequeuer   storage.LeafDequeuer
+}
+
+func (s logSequencingTask) load(ctx context.Context, limit int, cutoff time.Time) ([]*trillian.LogLeaf, error) {
+	start := s.timeSource.Now()
+	// Recent leaves inside the guard window will not be available for sequencing.
+	leaves, err := s.dequeuer.DequeueLeaves(ctx, limit, cutoff)
+	if err != nil {
+		glog.Warningf("%v: Sequencer failed to dequeue leaves: %v", s.label, err)
+		return nil, err
+	}
+	seqDequeueLatency.Observe(since(s.timeSource, start), s.label)
+
+	// Assign leaf sequence numbers.
+	for i, leaf := range leaves {
+		leaf.LeafIndex = s.treeSize + int64(i)
+	}
+	return leaves, nil
+}
+
+func (s logSequencingTask) store(ctx context.Context, leaves []*trillian.LogLeaf) error {
+	start := s.timeSource.Now()
+	// Write the new sequence numbers to the leaves in the DB.
+	if err := s.dequeuer.UpdateSequencedLeaves(ctx, leaves); err != nil {
+		glog.Warningf("%v: Sequencer failed to update sequenced leaves: %v", s.label, err)
+		return err
+	}
+	seqUpdateLeavesLatency.Observe(since(s.timeSource, start), s.label)
+	return nil
+}
+
+// IntegrateBatch wraps up all the operations needed to take a batch of queued
+// leaves and integrate them into the tree.
+//
+// TODO(Martin2112): Can possibly improve by deferring a function that attempts
```

Comment:
I think we can remove the todo as there are no current plans to do this and it's not clear that it would be better.

Reply:
Done.